### PR TITLE
NIFI-1831 Implemented encrypted configuration capabilities

### PR DESCRIPTION
**Pull request by alopresto**
This is a "beta" PR -- there are still some edge cases I want to clean up but I wanted to get eyes on it because it will be a large change (although most of the new code is in the unit tests). Specific areas of focus for review:
* LICENSE
* `AESSensitivePropertyProvider` -- class which actually performs encryption/decryption
* `ConfigEncryptionTool` -- command-line class to make the process of encrypting the configuration values easy

To be done:
* Assembly module to package command-line tool
* Resolve redundant logging dependencies in `nifi-encrypt-config`
* Handle `login-identity-providers.xml` properties
* Clean up `TODO`s
* Provide documentation for encrypted config values and tool usage
* Add new `EXIT_CODE`s for specific causes
* Add comments to some public methods

### CODE REVIEW

**Comment by alopresto** [comment]
I'm going to squash these initial commits as I don't believe anyone has started reviewing yet.

**Comment by joewitt** [comment]
The new dependency in NiFiProperties definitely starts to increase it's weight from a dependency/transitive perspective.  We could just tell them to only pay attention to its Properties interface but then they lose the static key name references and the convenience method calls.

The transitive dependency weight becomes even more problematic when we have additional providers beyond bouncycastle.

Perhaps we should consider ditching the static getInstance mechanism is NiFiProperties and instead have that class constructed at specific and appropriate entry points in the project.  The constructed instance would have populated plain text values for all properties sensitive or otherwise.  It could offer either an object model for properties which accounts for their key, value, and whether they're sensitive or not - OR - it could simply provide a method allowing the caller to ask if a given property was sensitive or not.  That said, I'm not sure why or if we'd want to expose whether a property was sensitive or not.  I could see it being useful we want to have callers know they better not store the value in a log file or exception message.

The ProcessContext (processors), Reporting Context (reporting tasks), and ConfigurtaionContext (controller services) each make sense for adding a 'getApplicationProperties()' method which returns the NiFiProperties.  This properties object would externalize any crypto libraries so that this dependency (nifi-properties) remains pure.

**Comment by alopresto** [comment]
Ok @joewitt . I agree that removing the singleton pattern for `NiFiProperties` is a good move. I will implement the loader (extracting the protection logic from `NiFiProperties`) and make an instance of the properties available for the various contexts. 

I will also create a mock loader implementation for the tests which reference `NiFiProperties`. 

I will deprecate `NiFiProperties.getInstance()` in `0.x` and remove it entirely in `1.0`.

**Comment by joewitt** [comment]
ok - love it.  My apologies for not noticing that earlier.  I'm happy to help with dividing and conquering cleanup on NiFiProperties usage.  This is something that has been a long time coming.  Your feature just tipped us over the edge of 'we better do something about this now'.  Thanks!

**Comment by alopresto** [comment]
@joewitt @mattyb149 @bbende @brosander 

As of right now, this PR (commit `ffab01d`) is rebased against the latest master. The tool is found in `nifi-toolkit/nifi-toolkit-assembly/target/nifi-toolkit-1.0.0-SNAPSHOT-bin/nifi-toolkit-1.0.0-SNAPSHOT/`. Running the command below (for example) will read an existing _nifi.properties_ file, encrypt all sensitive non-empty values using the provided _key_, populate those values (and the associated protection schemes -- `x.y.z.protected=aes/gcm/256`) into the new _nifi-encrypted.properties_ file, and persist the key in _bootstrap.conf_. 

* `$ ./bin/encrypt-config.sh -h` -- prints a usage message
* `$ ./bin/encrypt-config.sh -b path/to/bootstrap.conf -n path/to/nifi.properties -o path/to/nifi-encrypted.properties -p thisIsABadPropertiesPassword` -- normal use as described above
* `$ ./bin/encrypt-config.sh -b path/to/bootstrap.conf -n path/to/nifi.properties -o path/to/nifi-encrypted.properties -k 0123456789ABCDEFFEDCBA98765432100123456789ABCDEFFEDCBA9876543210` -- normal use as described above with raw hex key instead of password
* `$ ./bin/encrypt-config.sh -b path/to/bootstrap.conf -n path/to/nifi.properties -o path/to/nifi-encrypted.properties` -- normal use as described above but will prompt for key in secure console read

By default, it considers *sensitive* properties as anything that would be a password or key:
* `nifi.sensitive.props.key`
* `nifi.security.keystorePasswd`
* `nifi.security.keyPasswd`
* `nifi.security.truststorePasswd`

You can mark additional keys as *sensitive* by including them in a comma or semi-colon delimited string as follows (do this by hand in the input _nifi.properties_ before running the tool):
`nifi.sensitive.props.additional.keys=nifi.ui.banner.text`

Example:

*before* -- `~/Workspace/scratch/encrypted-configs/nifi.properties`
```
nifi.ui.banner.text=This is the banner text
...
# security properties #
nifi.sensitive.props.key=
nifi.sensitive.props.algorithm=PBEWITHMD5AND256BITAES-CBC-OPENSSL
nifi.sensitive.props.provider=BC
nifi.sensitive.props.additional.keys=nifi.ui.banner.text

nifi.security.keystore=keystore.jks
nifi.security.keystoreType=jks
nifi.security.keystorePasswd=thisIsABadKeystorePassword
nifi.security.keyPasswd=thisIsABadKeyPassword
nifi.security.truststore=truststore.jks
nifi.security.truststoreType=jks
nifi.security.truststorePasswd=thisIsABadTruststorePassword
nifi.security.needClientAuth=
nifi.security.user.authorizer=file-provider
nifi.security.user.login.identity.provider=
nifi.security.ocsp.responder.url=
nifi.security.ocsp.responder.certificate=
...
```

*run tool* -- 
```
hw12203:...assembly/target/nifi-toolkit-1.0.0-SNAPSHOT-bin/nifi-toolkit-1.0.0-SNAPSHOT (NIFI-1831) alopresto
🔓 167s @ 15:49:34 $ ./bin/encrypt-config.sh -b ~/Workspace/scratch/encrypted-configs/bootstrap.conf -n ~/Workspace/scratch/encrypted-configs/nifi.properties -o ~/Workspace/scratch/encrypted-configs/nifi-encrypted.properties -p thisIsABadPropertiesPassword
2016-08-19 15:57:48,097 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Invoked ConfigEncryptionTool with args [-b,/Users/alopresto/Workspace/scratch/encrypted-configs/bootstrap.conf,-n,/Users/alopresto/Workspace/scratch/encrypted-configs/nifi.properties,-o,/Users/alopresto/Workspace/scratch/encrypted-configs/nifi-encrypted.properties,-p,thisIsABadPropertiesPassword]
2016-08-19 15:57:48,794 INFO [main] o.a.nifi.properties.NiFiPropertiesLoader Loaded 112 properties from /Users/alopresto/Workspace/scratch/encrypted-configs/nifi.properties
2016-08-19 15:57:48,796 INFO [main] o.a.n.properties.ProtectedNiFiProperties Loaded 112 properties (including 0 protection schemes) into ProtectedNiFiProperties
2016-08-19 15:57:48,800 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Loaded NiFiProperties instance with 112 properties
2016-08-19 15:57:48,805 INFO [main] o.a.n.properties.ProtectedNiFiProperties Loaded 112 properties (including 0 protection schemes) into ProtectedNiFiProperties
2016-08-19 15:57:49,149 INFO [main] o.a.n.p.AESSensitivePropertyProvider AES Sensitive Property Provider encrypted a sensitive value successfully
2016-08-19 15:57:49,151 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Protected nifi.ui.banner.text with aes/gcm/256 -> 	2ZJZaFqqXl62HB5w||I57IDLE7hYJf2vJmrkC29ZjDztRJT00CVV1QkDiGte4VIfUB+n2X
2016-08-19 15:57:49,152 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Updated protection key nifi.ui.banner.text.protected
2016-08-19 15:57:49,152 INFO [main] o.a.n.p.AESSensitivePropertyProvider AES Sensitive Property Provider encrypted a sensitive value successfully
2016-08-19 15:57:49,152 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Protected nifi.security.keyPasswd with aes/gcm/256 ->     	icyABIxwN8FBBd6Q||GL+I6P5tv4ThiVXcAYkFaPsQLK2jZHyJui02W2YxydXfA0sEfg
2016-08-19 15:57:49,152 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Updated protection key nifi.security.keyPasswd.protected
2016-08-19 15:57:49,152 INFO [main] o.a.n.p.AESSensitivePropertyProvider AES Sensitive Property Provider encrypted a sensitive value successfully
2016-08-19 15:57:49,153 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Protected nifi.security.keystorePasswd with aes/gcm/256 ->        	eXpm2ylPEzfxpnEz||5Iqj7wReXHudmu/q7/q4vj2ooxFGN0R+o4VjV0BjboZ7zD6tqIEX3/Eq
2016-08-19 15:57:49,153 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Updated protection key nifi.security.keystorePasswd.protected
2016-08-19 15:57:49,153 INFO [main] o.a.n.p.AESSensitivePropertyProvider AES Sensitive Property Provider encrypted a sensitive value successfully
2016-08-19 15:57:49,153 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Protected nifi.security.truststorePasswd with aes/gcm/256 ->      	LGB8TUHtBTHemkzF||J5IqJpEB2i6QlFRqCEvgdq0koNPJv/oV4i3znyoAfzUV4laNv3TmWkeHFfs
2016-08-19 15:57:49,153 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Updated protection key nifi.security.truststorePasswd.protected
2016-08-19 15:57:49,167 INFO [main] o.a.n.properties.ProtectedNiFiProperties Loaded 115 properties (including 4 protection schemes) into ProtectedNiFiProperties
2016-08-19 15:57:49,167 INFO [main] o.a.nifi.properties.ConfigEncryptionTool Final result: 115 keys including 4 protected keys
2016-08-19 15:57:49,195 INFO [main] o.a.n.properties.ProtectedNiFiProperties Loaded 115 properties (including 4 protection schemes) into ProtectedNiFiProperties
hw12203:...assembly/target/nifi-toolkit-1.0.0-SNAPSHOT-bin/nifi-toolkit-1.0.0-SNAPSHOT (NIFI-1831) alopresto
🔓 495s @ 15:57:50 $
```

*after* -- `nifi-encrypted.properties`

```
nifi.ui.banner.text=2ZJZaFqqXl62HB5w||I57IDLE7hYJf2vJmrkC29ZjDztRJT00CVV1QkDiGte4VIfUB+n2X
nifi.ui.banner.text.protected=aes/gcm/256
...
# security properties #
nifi.sensitive.props.key=
nifi.sensitive.props.algorithm=PBEWITHMD5AND256BITAES-CBC-OPENSSL
nifi.sensitive.props.provider=BC
nifi.sensitive.props.additional.keys=nifi.ui.banner.text

nifi.security.keystore=keystore.jks
nifi.security.keystoreType=jks
nifi.security.keystorePasswd=eXpm2ylPEzfxpnEz||5Iqj7wReXHudmu/q7/q4vj2ooxFGN0R+o4VjV0BjboZ7zD6tqIEX3/Eq
nifi.security.keystorePasswd.protected=aes/gcm/256
nifi.security.keyPasswd=icyABIxwN8FBBd6Q||GL+I6P5tv4ThiVXcAYkFaPsQLK2jZHyJui02W2YxydXfA0sEfg
nifi.security.keyPasswd.protected=aes/gcm/256
nifi.security.truststore=truststore.jks
nifi.security.truststoreType=jks
nifi.security.truststorePasswd=LGB8TUHtBTHemkzF||J5IqJpEB2i6QlFRqCEvgdq0koNPJv/oV4i3znyoAfzUV4laNv3TmWkeHFfs
nifi.security.truststorePasswd.protected=aes/gcm/256
nifi.security.needClientAuth=
nifi.security.user.authorizer=file-provider
nifi.security.user.login.identity.provider=
nifi.security.ocsp.responder.url=
nifi.security.ocsp.responder.certificate=
```

I am still working on a couple small features like handling secure password entry to the console instead of just the key (and *then will remove the debugging of the invocation parameters which includes the key/password*). Please exercise what is here and report any issues. Thanks.

File: nifi-commons/nifi-properties/pom.xml [hunk 1]
```diff
@@ -21,4 +21,10 @@
         <version>1.0.0-SNAPSHOT</version>
     </parent>
     <artifactId>nifi-properties</artifactId>
+    <dependencies>
+        <dependency>
+            <groupId>org.bouncycastle</groupId>
+            <artifactId>bcprov-jdk15on</artifactId>
```
Comment (bbende) [inline nifi-commons/nifi-properties/pom.xml]: Should this be removed since we now have nifi-properties-loader in the framework which has all the bouncy castle stuff?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        setKey(key);
+    }
+
+    public AESSensitivePropertyProvider(byte[] key) throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException {
+        this(key == null ? "" : Hex.toHexString(key));
+    }
+
+    public void setKey(String keyHex) throws SensitivePropertyProtectionException {
+        if (!isKeySet()) {
+            if (keyHex == null || keyHex.length() == 0) {
```
Comment (bbende) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java]: May want to trim keyHex before checking length, or could use StringUtils.isBlank(keyHex)

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    private static final String ALGORITHM = "AES/GCM/NoPadding";
+    private static final String PROVIDER = "BC";
+    private static final String DELIMITER = "||"; // "|" is not a valid Base64 character, so ensured not to be present in cipher text
+    private static final int IV_LENGTH = 12;
+    private static final int MIN_CIPHER_TEXT_LENGTH = IV_LENGTH * 4 / 3 + DELIMITER.length() + 1;
+
+    private Cipher cipher;
+    private SecretKey key;
+
+    public AESSensitivePropertyProvider() {
```
Comment (bbende) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java]: Is there a use case for creating an instance of this class without an initial key and setting the key later? 
If there is then you can ignore this comment... Was thinking it might be simpler to make SecretKey final and only set from the constructor, and setKey could be private, and then no need to check if (!isKeySet()) when protecting/unprotecting

**Comment by bbende** [comment]
I ran into some test failures when building this PR, possibly due to the fact that I don't have JCE installed?

```
Failed tests:
  ConfigEncryptionToolTest.testShouldDeriveKeyFromPassword:337 assert derivedKey == KEY_HEX
       |          |  |
       |          |  0123456789ABCDEFFEDCBA98765432100123456789ABCDEFFEDCBA9876543210
       |          false
       0123456789ABCDEFFEDCBA9876543210
  ConfigEncryptionToolTest.testShouldPerformFullOperation Wrong exit status expected:<0> but was:<2>

Tests in error:
  ConfigEncryptionToolTest.testShouldEncryptSensitiveProperties:512 » IllegalArgument
  ConfigEncryptionToolTest.testShouldParseKey:279 » Key The key (64 hex chars) m...
  ConfigEncryptionToolTest.testShouldSerializeNiFiPropertiesAndPreserveFormat:751 » IllegalArgument
  ConfigEncryptionToolTest.testShouldSerializeNiFiPropertiesAndPreserveFormatWithNewPropertyAtEOF:798 » IllegalArgument
```

File: nifi-commons/nifi-properties/pom.xml (hunk 1 above)
Comment (alopresto) [inline nifi-commons/nifi-properties/pom.xml]: Good catch. Thanks Bryan.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java (hunk 2 above)
Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java]: Will do.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java (hunk 3 above)
Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/AESSensitivePropertyProvider.java]: There was a use case for this originally, but now the only access is from the constructor, so I will make this fix.

**Comment by alopresto** [comment]
Thanks Bryan. The tests were brittle and expected a 256-bit key in some cases. I have made adjustments to the test to be more robust in the absence of the JCE unlimited strength cryptography jurisdiction policy files.

**Comment by alopresto** [comment]
Thanks for all the comments @bbende . I have incorporated them and will be pushing a new commit as soon as the build & checkstyle complete.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/nifi.properties [hunk 4]
```diff
@@ -131,8 +131,10 @@ nifi.web.jetty.threads=${nifi.web.jetty.threads}
 
 # security properties #
 nifi.sensitive.props.key=
+nifi.sensitive.props.key.protected=${nifi.sensitive.props.key.protected}
```
Comment (bbende) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/nifi.properties]: After building this PR, in conf/nifi.properties the property nifi.sensitive.props.key.protected was empty. I then filled in a value for nifi.sensitive.props.key and when I ran the toolkit it did not protect nifi.sensitive.props.key, I assume because it saw nifi.sensitive.props.key.protected was already there? I tested it again by first deleting the empty protected property and that time it did protect the sensitive properties key.

Seems like nifi.sensitive.props.key.protected should not be here by default, or the toolkit should recognize an empty value and protect it.

File: nifi-toolkit/nifi-toolkit-encrypt-config/src/test/groovy/org/apache/nifi/properties/ConfigEncryptionToolTest.groovy [hunk 5]
```diff
@@ -0,0 +1,1225 @@
...
+        assert msg == "The nifi.properties file at ${workingFile.path} must be writable by the user running this tool".toString()
+
+        workingFile.deleteOnExit()
+    }
+
+    @Test
+    void testWriteNiFiPropertiesShouldHandleWriteFailureWhenFileDoesNotExist() {
+        // Arrange
+        File inputPropertiesFile = new File("src/test/resources/nifi_with_sensitive_properties_unprotected.properties")
+        File tmpDir = new File("tmp/")
```
Comment (bbende) [inline nifi-toolkit/nifi-toolkit-encrypt-config/src/test/groovy/org/apache/nifi/properties/ConfigEncryptionToolTest.groovy]: I think we should make all these references to "tmp/" be "target/tmp/", right now it leaves a "tmp" directory at the root of nifi-toolkit-encrypt-config after running a build

File: nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+
+    public ConfigEncryptionTool() {
+        this(DEFAULT_DESCRIPTION)
+    }
+
+    public ConfigEncryptionTool(String description) {
+        this.header = buildHeader(description)
+        this.options = new Options()
+        options.addOption("h", HELP_ARG, false, "Prints this usage message")
+        options.addOption("n", NIFI_PROPERTIES_ARG, true, "The nifi.properties file containing unprotected config values (will be overwritten)")
```
Comment (bbende) [inline nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy]: The first time I read the usage when I saw that -n said "will be overwritten", I assumed that if I didn't specify -o then it would just overwrite the file specified with -n, but it actually wrote to nifi-toolkit-1.0.0-SNAPSHOT/conf/nifi.properies, maybe just some additional wording?

**Comment by bbende** [comment]
General comment that may be outside the scope of this effort, but have we thought about being able to undo the protection using the toolkit? Basically once you have run the tool, if you overwrote your nifi.properties there is no real way to get back to the unprotected form.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/nifi.properties (hunk 4 above)
Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/nifi.properties]: Thanks for catching this Bryan. It was an issue where the tool had to merge two instances of `Properties` together and the logic wasn't checking the right key set. Added a test and fixed. Empty protection values no longer stop the field from being protected.

File: nifi-toolkit/nifi-toolkit-encrypt-config/src/test/groovy/org/apache/nifi/properties/ConfigEncryptionToolTest.groovy (hunk 5 above)
Comment (alopresto) [inline nifi-toolkit/nifi-toolkit-encrypt-config/src/test/groovy/org/apache/nifi/properties/ConfigEncryptionToolTest.groovy]: Meant to fix this the other day. Done.

File: nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy (hunk 6 above)
Comment (alopresto) [inline nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy]: The intended behavior is exactly as you originally described. I'm investigating this, but in general, if you run with `-n` but no `-o`, the file you specified is the one that should be overwritten.

Comment (bbende) [inline nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy]: It could be how I ran it... I was in a directory where I had the toolkit and nifi extracted, and I used a relative path like -n nifi-1.0.0-bin/conf/nifi.properties

**Comment by alopresto** [comment]
I have a note to possibly pause and prompt for confirmation on overwrite with the option to pass a flag allowing *force yes* behavior. I am hesitant to add "decryption" or "reversal" capability to the tool even requiring the password/key as input. 

Do you think what I suggested above would be sufficient? I am worried about someone trying to fire the tool without interaction and not realizing it will wait for confirmation, but I also accept that the tool is unlikely to be used in an unsupervised environment.

File: nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy (hunk 6 above)
Comment (alopresto) [inline nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy]: No, you were right. I added a test to capture it and resolved it.

**Comment by bbende** [comment]
I was actually thinking less about accidentally overwriting, and more about something like this... 

I setup nifi.properties, encrypt it, and everything is good, then I decide I want to add another encrypted property. Right now you can't run the tool a second time against an already protected nifi.properties, which makes sense, but it means you have to manually undo everything if you wanted to add another protected property.

I'm not hung up on this since it is not very hard to reset the file manually, just wanted to see what you thought about the reversal/decryption. I'm ok leaving everything as is for now.

**Comment by alopresto** [comment]
My intention is that you can run it repeatedly against the already-protected file if you add a new value. It should skip over the already protected values and only protect any new ones (marked as sensitive). Obviously you have to use the same password. Does this not work for you?

**Comment by bbende** [comment]
Ok that is not working for me... If I run the tool a second time using the same password, even without changing anything I get:
```
./nifi-toolkit-1.0.0-SNAPSHOT/bin/encrypt-config.sh -b nifi-1.0.0-SNAPSHOT/conf/bootstrap.conf -n nifi-1.0.0-SNAPSHOT/conf/nifi.properties
2016-08-23 14:56:51,772 WARN [main] o.a.nifi.properties.ConfigEncryptionTool The source nifi.properties and destination nifi.properties are identical [nifi-1.0.0-SNAPSHOT/conf/nifi.properties] so the original will be overwritten
Enter the password:
2016-08-23 14:56:56,604 INFO [main] o.a.nifi.properties.NiFiPropertiesLoader Loaded 118 properties from /Users/bbende/Projects/bbende-nifi/nifi-assembly/target/nifi-1.0.0-SNAPSHOT-bin/nifi-1.0.0-SNAPSHOT/conf/nifi.properties
2016-08-23 14:56:56,609 INFO [main] o.a.n.properties.ProtectedNiFiProperties Loaded 118 properties (including 3 protection schemes) into ProtectedNiFiProperties
Cannot load NiFiProperties from [nifi-1.0.0-SNAPSHOT/conf/nifi.properties]

usage: org.apache.nifi.properties.ConfigEncryptionTool [-b <arg>] [-h] [-k <arg>] [-n <arg>] [-o <arg>] [-p <arg>] [-r]
```

**Comment by bbende** [comment]
Running the latest commit, nifi.properties ends up with:
```
nifi.sensitive.props.key.protected=aes/gcm/128
nifi.sensitive.props.key.protected=
```
The app then fails to start because I think it finds the second one and can't decrypt the sensitive properties key. I can paste the stacktrace if helpful, but removing the empty one lets the app start.

**Comment by alopresto** [comment]
Ok I got it to a point where it can be run multiple times on an already-encrypted file and the only side-effect is that previously encrypted properties will look like (after 2 executions):

```
nifi.sensitive.props.key=C7AlglOx5QOmggfM||EG+qM9ubldyr8J73RmXCyWIxRKCzn6JyXv7mN6DkV51oDoRSdnWWXbALfKcKFw
nifi.sensitive.props.key.protected=aes/gcm/256
nifi.sensitive.props.key.protected=aes/gcm/256
```

Fixing that and then will push.

**Comment by bbende** [comment]
Performed a couple of test runs and everything is working nicely, thanks for making all the updates.

My last comment is that we may need to add something to the toolkit NOTICE for Grooby. The main NiFi assembly seems to have this for the scripting processors bringing in groovy, only difference being groovy.jar vs. groovy-all.jar: https://github.com/apache/nifi/blob/master/nifi-assembly/NOTICE#L815

If we add that (or we have a reason we don't need it) then I am a +1 and good to merge, nice work!

**Comment by alopresto** [comment]
I added *groovy-all* to `NOTICE` and added some Javadoc throughout the new classes. With Bryan's +1 I am going to squash the commits and merge to master.
